File: Java/descompositionLU.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class descompositionLU {
    private static final int N = 2000; 
    private static final double[][] M = new double[N][N];
    
    public static void main(String[] args) {
        LoadMatrix();
        //PrintMatrix(M, "M");
        System.out.println("Method Parallel");
        Parallel();
        System.out.println("Method Serial");
        Serial();
    }
    //=============================================================
    public static void LoadMatrix() {
        String fileName = "in.txt";
        readMatrixFromFile(fileName);
    }
    //=============================================================
    public static void readMatrixFromFile(String fileName) {
        try (Scanner scanner = new Scanner(new File(fileName))) {
            int rows = 0;
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                int k = line.length();
                if (k == 0) {
                    break;
                }
                String[] values = line.trim().split("\\s+");
                for (int col = 0; col < N; ++col) {
                    int temporal = Integer.parseInt(values[col]);
                    M[rows][col] = (double) temporal;
                }
                rows++;
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }
    //=============================================================
    public static void Serial() {
        //System.out.println("Descomposition LU in serial mode");
        long startTime = System.currentTimeMillis();
        DescompositionLUSerial(M);
        long endTime = System.currentTimeMillis();
        long totalTime = endTime - startTime;
        System.out.println("Execution Time: " + totalTime + " ms");
    }
    //=============================================================
    public static void Parallel() { 
        //System.out.println("Descomposition LU in parallel mode");
        long startTime = System.currentTimeMillis();
        DescompositionLUParallel(M);  
        long endTime = System.currentTimeMillis();
        long totalTime = endTime - startTime;
        System.out.println("Execution Time: " + totalTime + " ms");
    }
    //=============================================================
    public static void DescompositionLUSerial(double[][] M) {
        double[][] L = new double[N][N];
        double[][] U = new double[N][N];
        for (int i = 0; i < N; ++i) {
            L[i][i] = 1;
            for (int j = i; j < N; ++j) {
                double sum = 0;
                for (int k = 0; k < i; ++k) {
                    sum += L[i][k] * U[k][j];
                }
                U[i][j] = M[i][j] - sum; 
            }
            for (int j = i + 1; j < N; ++j) {
                double sum = 0;
                for (int k = 0; k < i; ++k) {
                    sum += L[j][k] * U[k][i];
                }
                L[j][i] = (M[j][i] - sum) / U[i][i];
            }
        }
        //PrintMatrix(L, "L");
        //PrintMatrix(U, "U");
    }
    //=================================================================
    public static void DescompositionLUParallel(double[][] M) {
        double[][] L = new double[N][N];
        double[][] U = new double[N][N];
        Thread[] threads = new Thread[8];
        final int[] current = {0};
        for (int i = 0; i < 8; ++i) {
            threads[i] = new Thread(() -> {
                while (true) {
                    int currentRow; 
                    synchronized (descompositionLU.class) {
                        if (current[0] >= N) {
                            break;
                        }
                        currentRow = current[0];
                        current[0]++;
                    }
                    L[currentRow][currentRow] = 1;
                    for (int j = currentRow; j < N; ++j) {
                        double sum = 0;
                        for (int k = 0; k < currentRow; ++k) {
                            sum += L[currentRow][k] * U[k][j];
                        }
                        U[currentRow][j] = M[currentRow][j] - sum;
                    }
                    for (int j = currentRow + 1; j < N; ++j) {
                        double sum = 0;
                        for (int k = 0; k < currentRow; ++k) {
                            sum += L[j][k] * U[k][currentRow];
                        }
                        L[j][currentRow] = (M[j][currentRow] - sum) / U[currentRow][currentRow];
                    }
                }
                
            });
            threads[i].start();
        }
        for (Thread thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        //PrintMatrix(L, "L");
        //PrintMatrix(U, "U");
    }
    //=================================================================
    public static void PrintMatrix(double[][] X, String NAME) {
        System.out.println("Matrix " + NAME + ":");
        for (double[] row : X) {
            for (double value : row) {
                System.out.print(value + "\t");
            }
            System.out.println();
        }
        System.out.println();
    }
}

File: KotlinTest/descompositionLU.kt
//import kotlin.ExperimentalStdlibApi
import kotlin.random.Random
import kotlin.concurrent.thread

val N = 3000
val M = Array(N) {DoubleArray(N)}

fun main() {
    loadMatrix();
    println("Method Parallel:")
    parallel()
    println("Method Serial:")
    serial()
}
//===================================================
fun loadMatrix() : Unit {
    for (i in 0 until N) {
        for (j in 0 until N) {
            var e = Random.nextInt(1, 11)
            M[i][j] = e.toDouble()
        }
    }
}
//===================================================
fun serial() : Unit {
    //println("Descomposition LU in serial mode")
    val start = System.currentTimeMillis()
    descompositionLUSerial(M)
    val end = System.currentTimeMillis()
    val total = end - start
    println("Execution Time: " + total + " ms")
} 
//===================================================
fun parallel() : Unit {
    //println ("Descomposition LU in parallel mode")
    val start = System.currentTimeMillis()
    descompositionLUParallel(M)
    val end = System.currentTimeMillis()
    val total = end - start
    println("Execution Time: " + total + " ms")
}
//===================================================
fun descompositionLUSerial(M: Array<DoubleArray>) : Unit {
    val L = Array(N) {DoubleArray(N)}
    val U = Array(N) {DoubleArray(N)}
    for (i in 0 until N) {
        L[i][i] = 1.0;
        for (j in i until N) {
            var sum = 0.0
            for (k in 0 until i) {
                sum += L[i][k] * U[k][j]
            }
            U[i][j] = M[i][j] - sum
        }
        for (j in i + 1 until N) {
            var sum = 0.0
            for (k in 0 until i) {
                sum += L[j][k] * U[k][i]
            }
            L[j][i] = (M[j][i] - sum) / U[i][i]
        }
    }
    //printMatrix(L, "L")
    //printMatrix(U, "U")
}
//===================================================
//@OptIn(ExperimentalStdlibApi::class)
fun descompositionLUParallel(M: Array<DoubleArray>) : Unit {
    val L = Array(N) {DoubleArray(N)}
    val U = Array(N) {DoubleArray(N)}
    val threads = arrayOfNulls<Thread>(8)
    val current = intArrayOf(0)

    for (i in 0 until 8) {
        threads[i] = thread(start = true) {
            while (true) {
                val currentRow: Int
                synchronized(Any()) {
                    if (current[0] >= N) {
                        return@thread
                    }
                    currentRow = current[0]
                    current[0]++
                }
                L[currentRow][currentRow] = 1.0
                for (j in currentRow until N) {
                    var sum: Double = 0.0
                    for (k in 0 until currentRow) {
                        sum += L[currentRow][k] * U[k][j]
                    }
                    U[currentRow][j] = M[currentRow][j] - sum
                }
                for (j in currentRow + 1 until N) {
                    var sum: Double = 0.0
                    for (k in 0 until currentRow) {
                        sum += L[j][k] * U[k][currentRow]
                    }
                    L[j][currentRow] = (M[j][currentRow] - sum) / U[currentRow][currentRow]
                }
            }

        }
    }
    threads.forEach { it?.join()}
    //printMatrix(L, "L")
    //printMatrix(U, "U")
}
//===================================================
fun printMatrix(M: Array<DoubleArray>, NAME: String) : Unit {
    println("Matrix $NAME")
    for (row in M) {
        for (value in row) {
            print("$value\t")
        }
        println()
    }
    println()
}
//===================================================

